src/f32-igemm/gen/f32-igemm-7x4v-minmax-rvv.c

```diff
@@ -0,0 +1,183 @@
+  assert(c != NULL);
+
+  const float vmin = params->scalar.min;
+  const float vmax = params->scalar.max;
+  float* c0 = c;
+  float* c1 = (float*) ((uintptr_t) c0 + cm_stride);
+  if XNN_UNPREDICTABLE(mr < 2) {
+    c1 = c0;
+  }
+  float* c2 = (float*) ((uintptr_t) c1 + cm_stride);
+  if XNN_UNPREDICTABLE(mr <= 2) {
+    c2 = c1;
+  }
+  float* c3 = (float*) ((uintptr_t) c2 + cm_stride);
+  if XNN_UNPREDICTABLE(mr < 4) {
+    c3 = c2;
+  }
+  float* c4 = (float*) ((uintptr_t) c3 + cm_stride);
+  if XNN_UNPREDICTABLE(mr <= 4) {
+    c4 = c3;
+  }
+  float* c5 = (float*) ((uintptr_t) c4 + cm_stride);
+  if XNN_UNPREDICTABLE(mr < 6) {
+    c5 = c4;
+  }
+  float* c6 = (float*) ((uintptr_t) c5 + cm_stride);
+  if XNN_UNPREDICTABLE(mr <= 6) {
+    c6 = c5;
+  }
+
+  const size_t nr = __riscv_vsetvlmax_e32m4();
+  size_t vl = nr;
+  do {
+    if XNN_UNLIKELY(nc < nr) {
+      vl = __riscv_vsetvl_e32m4(nc);
+    }
+    nc = nc - vl;
+    vfloat32m4_t vacc0 =  __riscv_vle32_v_f32m4(w, vl);
+    w = w + nr;
+    vfloat32m4_t vacc1 =  __riscv_vmv_v_v_f32m4(vacc0, vl);
+    vfloat32m4_t vacc2 =  __riscv_vmv_v_v_f32m4(vacc0, vl);
+    vfloat32m4_t vacc3 =  __riscv_vmv_v_v_f32m4(vacc0, vl);
+    vfloat32m4_t vacc4 =  __riscv_vmv_v_v_f32m4(vacc0, vl);
+    vfloat32m4_t vacc5 =  __riscv_vmv_v_v_f32m4(vacc0, vl);
+    vfloat32m4_t vacc6 =  __riscv_vmv_v_v_f32m4(vacc0, vl);
+
+    size_t p = ks;
+    do {
+      const float* restrict a0 = a[0];
+      assert(a0 != NULL);
+      if XNN_UNPREDICTABLE(a0 != zero) {
+        a0 = (const float*) ((uintptr_t) a0 + a_offset);
+      }
+      const float* restrict a1 = a[1];
+      assert(a1 != NULL);
+      if XNN_UNPREDICTABLE(a1 != zero) {
+        a1 = (const float*) ((uintptr_t) a1 + a_offset);
+      }
+      const float* restrict a2 = a[2];
+      assert(a2 != NULL);
+      if XNN_UNPREDICTABLE(a2 != zero) {
+        a2 = (const float*) ((uintptr_t) a2 + a_offset);
+      }
+      const float* restrict a3 = a[3];
+      assert(a3 != NULL);
+      if XNN_UNPREDICTABLE(a3 != zero) {
+        a3 = (const float*) ((uintptr_t) a3 + a_offset);
+      }
+      const float* restrict a4 = a[4];
+      assert(a4 != NULL);
+      if XNN_UNPREDICTABLE(a4 != zero) {
+        a4 = (const float*) ((uintptr_t) a4 + a_offset);
+      }
+      const float* restrict a5 = a[5];
+      assert(a5 != NULL);
+      if XNN_UNPREDICTABLE(a5 != zero) {
+        a5 = (const float*) ((uintptr_t) a5 + a_offset);
+      }
+      const float* restrict a6 = a[6];
+      assert(a6 != NULL);
+      if XNN_UNPREDICTABLE(a6 != zero) {
+        a6 = (const float*) ((uintptr_t) a6 + a_offset);
+      }
+      a += 7;
+
+      size_t k = kc;
+      do {
+        const float va0 = *a0++;
+        const float va1 = *a1++;
+        const float va2 = *a2++;
+        const float va3 = *a3++;
+        const float va4 = *a4++;
+        const float va5 = *a5++;
+        const float va6 = *a6++;
+        vfloat32m4_t vb = __riscv_vle32_v_f32m4(w, vl);
```

Comment:
overall this looks awesome.  But I wonder if this load is a bottleneck and should be scheduled carefully.
in arm assembly kernels I would preload some of the weights and inputs in a prologue, so in the loop the first instruction is a vfmacc, and then co-issue loads for the next iteration of the loop.
It usually starts by unrolling the loop to do 2 iterations, so you can load the weights for the next iteration while vfmacc-ing the current weights.
And then a prologue that is the same as the main loop, but no loads for the next iteration.
Also on larger models (LLMs) I'm finding a prefetch on weights helps.   That might be hard for RVV, not knowing how many bytes you're loading, so an alternative would be a redundent vle32 far into the future weights, but ignore the result?  The distance is usually fixed in practoce - 448 for arm and 960 for intel, depending on your cache.  448 is 7 cache lines of 64 for an 8 way cache.  960 is for a 15 cache lines of 64, for a 16 way cache.  Going beyond those sizes can cause collisions and performance can drop off.